src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseStream.cs

```diff
@@ -227,5 +243,33 @@ private void CheckDisposed()
                 throw new ObjectDisposedException(this.GetType().FullName);
             }
         }
+        
+        // The only way to abort pending async operations in WinHTTP is to close the request handle.
+        // This causes WinHTTP to cancel any pending I/O and accelerating its callbacks on the handle.
+        // This causes our related TaskCompletionSource objects to move to a terminal state.
+        //
+        // We only want to dispose the handle if we are actually waiting for a pending WinHTTP I/O to complete,
+        // meaning that we are await'ing for a Task to complete. While we could simply call dispose without
+        // a pending operation, it would cause random failures in the other threads when we expect a valid handle.
+        private void CancelPendingResponseStreamReadOperation()
```

Comment:
/cc @stephentoub 
That's not the scenario I was referring to: I thought that the cancellation callback can be called from a timer thread. 
If that's true, it means that it can be called even after `DisposeCtrReadFromResponseStream()` started execution (there's nothing to synchronize the two). If that happens, the code ends up with the `_requestHandle` disposed while executing the WinHttp callback code.